### ISSUE TITLE
Update SIP coefficient accuracy

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#8341]
_Issue [JP-3567](https://jira.stsci.edu/browse/JP-3567) was created on JIRA by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw):_

A common issue that users encounter for large imaging is WCS imprecision arising from too-few coefficients in the FITS SIP representation.  Ideally, users would use the ASDF metadata instead, as this is a more-accurate representation of the WCS in all modes.  However, we can nonetheless improve the accuracy of the SIP coefficients as well.

This would require increasing the polynomial order (potentially from 3 to 5), and also updating the polynomials whenever the ASDF WCS information is updated (e.g., by tweakreg).  This should apply to the primary pipeline products, as well as things like the crf.fits files.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=688010#comment-688010):_

[Armin Rest](https://jira.stsci.edu/secure/ViewProfile.jspa?name=arest) Would you be able to provide more information for this ticket as necessary?

**Comment by stscijgbot-jp** [comment]
_Comment by [Armin Rest](https://jira.stsci.edu/secure/ViewProfile.jspa?name=arest) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=688720#comment-688720):_

I support increasing the SIP polynomial order to 5. I had cases in which re-calculating the SIP solution after the asdf was updated failed, which is very inconvenient. I don't think there is the danger of overfitting and getting bad values at the edges or corner, since the data to be fitted is not sparse by design (I assume a fine grid of values is fitted).

I also support that SIP is automatically updated whenever the asdf solution is changed, e.g. by tweakreg. Right now, this is a two step process, it would be nice to have this done automatically....

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=689824#comment-689824):_

Currently the code iterates over a list of degrees until the fit achieves a pixel error of 0.01. If you supply a degree there's no guarantee the solution will converge for this degree or that it won't overfit. I have two questions and the solution will be different based on the answer.
 * Do you find the actual pixel error after running thе approximation is greater than 0.01?
 * Or do you think 0.01 pixel error is not of sufficient accuracy?

Of course all these parameters are settable by the user. The question is what would work best in an automated environment.

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=689825#comment-689825):_

Also the tweakreg code updates the SIP coefficients. Do you find this is not the case?

**Comment by stscijgbot-jp** [comment]
_Comment by [Armin Rest](https://jira.stsci.edu/secure/ViewProfile.jspa?name=arest) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=689842#comment-689842):_

At some point in the past this was a two-step process, i.e. tweakreg did not update the SIP coefficients, and we have to do an extra call to update the SIP coefficients. [Justin Pierel](https://jira.stsci.edu/secure/ViewProfile.jspa?name=jpierel) might be able to comment here...

**Comment by drlaw1558** [comment]
@jdavies-st may also be able to chime in, as he'd been finding that the crf files contained the tweaked gwcs but the old FITS SIP approximations.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=689948#comment-689948):_

It would also be useful for [Varun Bajaj](https://jira.stsci.edu/secure/ViewProfile.jspa?name=vbajaj) to chime in, as he had done some testing of this for the workshop back in November.

**Comment by Vb2341** [comment]
I just took a quick look into this again, so I'm not sure if I've got it exactly pinned down.  That said, I believe the issue is actually just that the call to `update_fits_wcsinfo` in the assign_wcs step ends up using the default value of 0.1 for `sip_max_pix_error`.   I think the value of `sip_degree` probably also uses the value of `None`, so then the polynomial degree is determined by `sip_max_pix_error`, and (for NIRCam at least) that requires in a SIP degree of 3 to achieve the <0.1 pix error.  The same behavior is true for the inverse polynomials, I think.  For the one nircam image i quickly looked at, this gets max errors of about 0.04 pixels, which is significant.

However, the call to `update_fits_wcsinfo` in the tweakregstep specifies a `sip_max_pix_error` of 0.01.  In the one test case I looked at, this results in a SIP degree of 5.  It doesn't specify anything for the inverse transform however, so I assume it would use the same 0.1 threshold, but the inverse polynomial keywords don't appear in the fits header, so I'm not too sure what's really happening there (the asdf extension has values of None for all the inverse polynomials in the tweakregstep output files).  

In summary- I think the problem is in the default arguments of the assign_wcs step (0.1 pixel threshold).  I'm not sure whether it's appropriate to update the default values in the step itself, or if this is a change to be made in a reference file somewhere.

**Comment by jdavies-st** [comment]
Yeah, this may be the case where for NIRCam detectors the defaults need to be set in an `assign_wcs` param reffile, but really, whatever is being done in the initial `assign_wcs` and in `tweakreg` needs to be consistent.  So number of points used, tolerance, etc.  So those parameters need to be coordinated as either defaults for all instruments or defaults that both steps can use.

The reason is important - only some of the images get modified when `tweakreg` runs.  The reference image (almost always the first in the association) does not, which means it retains the original -SIP approximation from `assign_wcs`.  So whatever parameters are appropriate, `assign_wcs` and `tweakreg` need to be in lockstep.

Here's an illustration of the problem using some F200W and F470N data from program ID 2321.  Standard jwst 1.14.0 processing using defaults with `assign_wcs`.

```python
In [1]: from glob import glob
   ...: from jwst import datamodels
   ...: from gwcs.wcstools import grid_from_bounding_box
   ...: from astropy.wcs import WCS
   ...: from astropy.io import fits
   ...: import numpy as np
   ...: from matplotlib import pyplot as plt

In [131]: cal_files = glob("jw02321001007_02101_00001_nrc*_cal.fits")
     ...: cal_files.sort()
     ...: fig, axs = plt.subplots(2, 5)
     ...: for i, f in enumerate(cal_files):
     ...:     model = datamodels.open(f)
     ...: 
     ...:     # What is the order of the -SIP approximation?
     ...:     a_order = model.meta.wcsinfo.a_order
     ...: 
     ...:     # Make a grid of x,y pixel coordinates to pass to WCSs
     ...:     grid = grid_from_bounding_box(model.meta.wcs.bounding_box)
     ...: 
     ...:     # Compute GWCS world coords
     ...:     world_g = model.meta.wcs.pixel_to_world(*grid)
     ...:     hdulist = fits.open(f)
     ...:     w = WCS(hdulist['SCI'].header)
     ...: 
     ...:     # Compute FITS -SIP world coords
     ...:     world_f = w.pixel_to_world(*grid)
     ...: 
     ...:     # The angular separation between GWCS and -SIP approximation
     ...:     sep = world_g.separation(world_f)
     ...: 
     ...:     # Plot it all up
     ...:     detector = f[26:].split('_')[0]
     ...:     pixel_scale = 0.031
     ...:     if "long" in detector:
     ...:         pixel_scale = 0.063
     ...:     sep_in_pix = sep.arcsec / pixel_scale
     ...: 
     ...:     # Compute percentage of pixels over tolerance assuming average pixel scale
     ...:     percent_over_tolerance = 100 * np.count_nonzero(sep_in_pix > 0.01) / 2048**2
     ...: 
     ...:     cs = axs.flatten()[i].imshow(sep_in_pix, vmin=0, vmax=0.05)
     ...:     if i % 5 == 4:
     ...:         fig.colorbar(cs)
     ...:     axs.flatten()[i].set_title(f"{detector.upper()}   a_order={a_order}  {percent_over_tolerance:4.1f}%")
     ...: plt.show()
```

<img width="2261" alt="Screenshot 2024-05-22 at 12 12 04" src="https://github.com/spacetelescope/jwst/assets/17088869/d8fc7c86-2b97-456a-880a-909aa3057705">

I've plotted them all on the same scale above, and as you can see, the ones that have order=3 are much worse than order=4, but many have a significant percentage of pixels in the pixel grid over a tolerance of 0.01 pixels.

This is just for the case of F200W and F470N of this program. Since the distortion model for each detector/filter combo is different, you'll probably see different behavior for other filters.

And while setting a tolerance is important, also a well-sampled grid over which that tolerance is evaluated is important too.  Right now, the defaults are a 12x12 grid for `assign_wcs` and a 32x32 grid for `tweakreg`.  That's every 171 pixels and every 64 pixels, respectively.  If we expect there to be wiggles on a scale that is smaller than this, then perhaps look at a finer grid?  Certainly the grids should be the same for both steps.

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=700363#comment-700363):_

Yes, the degree of the polynomial is controlled by the requested pixel error. I'll double check all defaults.

**Comment by jdavies-st** [comment]
I would also argue that the vast majority of the data for which alignment, tweakreg and resampling is important is NIRCam data, so defaults that work for NIRCam data (and are not bad for NIRISS or MIRI) should be implemented.  Obviously one would want to carry out the exercise I did above with various methods of running the -SIP approximation for all detector/filter combos and see what works for all. It may be that hard-coding it to a specific order instead of the default of specifying a tolerance and then increasing order iteratively until the tolerance is met.  I.e. specify order=5 if needed and do the fit once.

Theoretically, the distortion models do not change for a given detector/filter/pupil combo.  You could compute them once and be done with it.  The only caveat being the affine transformation that `tweakreg` applies when improving alignment, but this is essentially modifying the linear terms, not the non-linear terms.

I'm not arguing for putting the -SIP approximation in the distortion reference file, but I am saying that once you figure out what order polynomial on what grid gets you to the desired precision for each detector/filter/pupil combo, then that should work every time as its the exact same computation for the non-linear terms, so there's no need to do an iterative approach as one would do in the general case.

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=701770#comment-701770):_

It sounds like this needs a few fixes on the software side:
1. Update the default parameters in assign_wcs to match the default values for the update_fits_wcs function, so that max_pix_error is 0.01 everywhere and npoints is 32 everywhere.
2. Explicitly specify the inverse parameter defaults in update_fits_wcs and assign them to the same defaults as the forward parameters.
3. Add step parameters to tweakreg, matching the parameters in assign_wcs, so the later WCS updates match the initial ones.

Then, instrument teams can decide if they would like to make parameter reference files to override the defaults for assign_wcs and tweakreg.  If they do, they should change both sets at the same time.

Does that sound right?

**Comment by stscijgbot-jp** [comment]
_Comment by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=701787#comment-701787):_

[Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) I think this is a good plan.

**Comment by melanieclarke** [comment]
Thanks @jdavies-st for the testing code and data.  I was able to reproduce your plot, above, with the current main branch. Here's the equivalent plot for the same cal files after image2 with my PR branch:

![image_comparison](https://github.com/spacetelescope/jwst/assets/2838388/5ec7578a-2c7a-45b0-aee4-ac3a2dadb3f1)

All pixels are well within the 0.01 tolerance, closer to ~1e-6, with the new default parameters.

I was also able to reproduce the other effect you mentioned, where output SIP approximations from tweakreg had order 3-4 for the first image, but 5 for the others.  With these changes, matching tweakreg to assign_wcs, order = 5 for all output from tweakreg.

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=702884#comment-702884):_

Reviewing regression tests for the PR, [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) noted an unexpected change for a MIRI image.  It turns out that with the new defaults for assign_wcs, the SIP approximation fails to converge for this MIRI image to the desired accuracy.

I have not tested a range of data, so I'm not sure how frequently this might happen, but I'm wondering if we should keep the defaults as is in the assign_wcs step (and set the tweakreg defaults to match) to avoid degrading the fit for MIRI.  Then, if other instruments need higher accuracy, they can submit parameter reference files to override the defaults.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=702902#comment-702902):_

Hm, one challenge with that is that there are currently no pars-assignwcsstep param reference files defined in CRDS, which would require a code change to enable.  We're getting a little down to the wire on enabling that in 11.0.  Is it obvious what's going on with the MIRI image to cause the convergence failure?

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=702907#comment-702907):_

It certainly could be useful to test the updates on a few more MIRI images than just the one in the regression test suite that showed poor results (the fit could not converge using the higher polynomial order). On the other hand, assuming most MIRI images have approximately the same kinds of distortion terms, which are independent of anything like image content, I'm guessing the fits would run into similar issues in most/all images.

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=702928#comment-702928):_

Can you remind me what happens when the SIP approximation doesn't converge to the desired accuracy?  Does this cause a crash or just a warning that the final fit isn't as good as requested?  I.e., are we actually making the MIRI data worse or just not making them better?

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=702965#comment-702965):_

Too bad it can't do something like attempting to fit with the requested npoints and if it doesn't converge, automatically back off on the order until it does converge.

**Comment by stscijgbot-jp** [comment]
_Comment by [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=702974#comment-702974):_

It seems to be npoints = 32 that is the sticking point for MIRI.  With npoints = 12 (old default), for the test image (jw01024001001_04101_00001_mirimage_cal.fits) it warns that it can't get to the desired output accuracy (0.01 pixels), but the output WCS is still pretty good (final order 4 or 5, diffs from the real WCS up to ~1e-6 pixels).  With npoints = 32, it just fails to find a fit at all (final order 2, all coefficients 0.0, diffs from the real WCS up to ~14 pixels).

I checked with the NIRCam data set what it looks like with npoints = 12 and pixel accuracy 0.01, and that still looks pretty good. The differences from using npoints = 32 are very minor, with diffs from the real WCS ~1e-7 in both cases.

How about I set the default npoints back to 12 but leave the accuracy parameters at 0.01?

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=703220#comment-703220):_

[Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) Sounds good to me.

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=703237#comment-703237):_

Fixed by #8529

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3567?focusedCommentId=707109#comment-707109):_

Tested that I get the same results for James' test example that [Melanie Clarke](https://jira.stsci.edu/secure/ViewProfile.jspa?name=mclarke) did, with the tagged 1.15.0 code.  Closing.
